### PR TITLE
Add "the tartan universe"

### PR DESCRIPTION
**Pull request by kaalleen**
For more details see https://inkstitch.org/docs/stitches/tartan-fill/

### CODE REVIEW

**Comment by lexelby** [comment]
Holy crap that's amazing!!  I don't quite understand how tartans work, but this is really fun to play with even still.

I got this error trying out the "SVG Elements" option:

```
Traceback (most recent call last):
  File "/home/lex/repos/inkstitch/lib/gui/tartan/main_panel.py", line 207, in apply
    self.update_tartan()
  File "/home/lex/repos/inkstitch/lib/gui/tartan/main_panel.py", line 264, in update_tartan
    TartanSvgGroup(self.settings).generate(element)
  File "/home/lex/repos/inkstitch/lib/tartan/svg.py", line 115, in generate
    fills, strokes = self._combine_shapes(warp, weft, outline_shape)
  File "/home/lex/repos/inkstitch/lib/tartan/svg.py", line 454, in _combine_shapes
    travel_path_element = self._linestring_to_path(color, travel, shapes[0][0].get('transform'), True)
  File "/home/lex/repos/inkstitch/lib/tartan/svg.py", line 583, in _linestring_to_path
    path_element = PathElement(
  File "src/lxml/classlookup.pxi", line 77, in lxml.etree.ElementBase.__init__
  File "src/lxml/apihelpers.pxi", line 158, in lxml.etree._initNewElement
  File "src/lxml/apihelpers.pxi", line 318, in lxml.etree._initNodeAttributes
  File "src/lxml/apihelpers.pxi", line 335, in lxml.etree._addAttributeToNode
  File "src/lxml/apihelpers.pxi", line 1539, in lxml.etree._utf8
TypeError: Argument must be bytes or unicode, got 'NoneType'
```

**Comment by lexelby** [comment]
That was with this tartan:

```
{"symmetry": true, "equal_warp_weft": false, "rotate": 0.0, "scale": 100, "offset_x": 0.0, "offset_y": 0.0, "pallet": "(#000000)/3.0 (#F6D32D)?5.0 (#33D17A)7.0 | (#F6D32D)?5.0 (#000000)8.0 (#33D17A)/7.0", "output": "embroidery", "stitch_type": "auto_fill", "row_spacing": 1.0, "angle_warp": 0.0, "angle_weft": 90.0, "min_stripe_width": 1.0, "bean_stitch_repeats": 0}
```

File: lib/stitches/auto_fill.py [hunk 1]
```diff
@@ -57,7 +57,7 @@ def is_outline(self):
         return self.key in self.OUTLINE_KEYS
 
     def is_segment(self):
-        return self.key == self.SEGMENT_KEY
+        return self.key.startswith(self.SEGMENT_KEY)
```
Comment (lexelby) [inline lib/stitches/auto_fill.py:60]: Do we need to do similar for `is_outline()`?

File: lib/tartan/pallet.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+import re
+from typing import List
+
+import wx
+from inkex import Color
+
+from .colors import string_to_color
+
+
+class Pallet:
```
Comment (lexelby) [inline lib/tartan/pallet.py:16]: English nit: we want to spell this as "palette", otherwise we're talking about a bed :)

**Comment by lexelby** [comment]
I added a couple comments, but this looks great to me!  I think I would probably need a class in how tartans work before I could understand the tartan fill code itself, but that's fine.  I like that you're using typing, definitely something we should do as much as possible. :+1:

**Comment by kaalleen** [comment]
> I got this error trying out the "SVG Elements" option:

I cannot reproduce. Do you still have the svg file?

File: lib/tartan/pallet.py (hunk 2 above)
Comment (kaalleen) [reply lib/tartan/pallet.py:16]: Oh no. You read my real thoughts while writing that, let's wake up then :)

File: lib/stitches/auto_fill.py (hunk 1 above)
Comment (kaalleen) [reply lib/stitches/auto_fill.py:60]: Yes, luckily startswith works with tuples as well.

**Comment by kaalleen** [comment]
Ok, I figured it out. It happens when you have a document which is already in px units which lead to a None type transform.

**Pull request merged**